astroquery/mast/observations.py

```diff
@@ -610,9 +613,12 @@ def _download_files(self, products, base_dir, *, cache=True, cloud_only=False,):
         for data_product in products:
 
             # create the local file download path
-            local_path = os.path.join(base_dir, data_product['obs_collection'], data_product['obs_id'])
-            if not os.path.exists(local_path):
-                os.makedirs(local_path)
+            if not flat:
+                local_path = os.path.join(base_dir, data_product['obs_collection'], data_product['obs_id'])
+                if not os.path.exists(local_path):
+                    os.makedirs(local_path)
+            else:
+                local_path = base_dir
```

Comment:
Have you considered using `Path` from [`pathlib`](https://docs.python.org/3.7/library/pathlib.html)? 
```suggestion
            if flat:
                local_path = Path(base_dir)
            else:
                local_path = Path(base_dir, data_product['obs_collection'], data_product['obs_id'])
                local_path.mkdir(parents=True, exist_ok=True)
```

Reply:
I did consider, I just didn't know what the downstream effects of that would be, given that all paths in the module use the older `os.path` routines, as do most in the core library.  You'll notice in the test I did use the pytest `tmp_path` fixture, which supplies a temporary `Path` object as location to write the temporary downloaded file.  And it works just fine.  So it should work, but maybe it would be good to do a consistent cleanup throughout the module?

I did fix

https://github.com/astropy/astroquery/pull/2511/files#diff-55036b43100098d3295a04581bde036bd91d0371ed055fb0dabd0d8ddbfb6b5bL738

```python
base_dir = download_dir.rstrip('/') + "/mastDownload"
```

which I believe was probably a bug, as that cannot work with a Windows filesystem.  But I held off doing anymore changes that were not focused on adding this new feature.  Happy to do that in a subsequent PR so as not to detract from this one.

Reply:
Yes, it would be better to be consistent throughout the module and switching to `pathlib` sounds like something that deserves a separate pull request.

Reply:
I would rather not mix the two without a good reason, and even `pathlib`'s own documentation refers that `os.path` can be used. That being said I would merge a PR if it consistently changes all os.path to pathlib in the package (but if so, I would still not like to see `/` operators as those may be misleading during quick code reads (and because we still get a lot of contributions doing raw sting manipulations and additions using `'/'` in them.